### ISSUE TITLE
Investigate alternative algorithms to RSA for webhook communication

### ISSUE THREAD

**Issue by c-pius** [kyma-project/lifecycle-manager#1795]
### Description

Follow up from https://github.com/kyma-project/lifecycle-manager/issues/1771

We want to investigate if there are alternative algorithms to RSA for our webhoook communication.

### Reasons

RSA seems to be not state of the art any longer (see https://github.com/kyma-project/lifecycle-manager/issues/1771#issuecomment-2296230848). In addition, we observed problems when testing RSA and 8192 bit key size: https://github.com/kyma-project/lifecycle-manager/issues/1793

### Acceptance Criteria

- [x] investigated alternatives to RSA algorithm that are compliant with https://github.tools.sap/kyma/backlog/issues/5825
- [x] understood and documented the tradeoffs when switching to a new algorithm (e.g., security, performance, ...)
- [x] decided whether to stick with RSA or update to a new algorithm with the team and created necessary follow-ups
  - [ ] if changing to a new algorithm, https://github.com/kyma-project/lifecycle-manager/issues/1793 can be closed

Timebox: 2 days (consider small PoC)

**Comment by c-pius** [comment]
> - [ ] investigated alternatives to RSA algorithm that are compliant with https://github.tools.sap/kyma/backlog/issues/5825
> - [ ] understood and documented the tradeoffs when switching to a new algorithm (e.g., security, performance, ...)
> 

### RSA

`RSA` still seems to be the most popular algorithm for creating key pairs. What has been deprecated is the `SHA1` hashing algorithm for signing the certificate. The secure successor to use is `SHA2`. `SHA2` refers to a family of hasing algorithms where `SHA-2 == SHA-256` and `SHA-224`, `SHA-384` and `SHA-512` are referring to alternative bit lengths of `SHA-2`.

Looking at our watcher certificates: `kubectl get secret skr-webhook-tls -o yaml -n kyma-system | yq '.data."tls.crt"' | base64 -d | openssl x509 -noout -text`, we can see `Signature Algorithm: sha256WithRSAEncryption` which is up to date.

cert-manager derives the signing algorithm from the key length and always uses SHA-2. See [func SignatureAlgorithm](https://github.com/cert-manager/cert-manager/blob/v1.15.0/pkg/util/pki/csr.go#L450-L464) (1.15.0).

Product standard security (SEC-266) further states:

- Recommended are RSA with a key length beyond 3000 bit
- Recommended hashing algorithms for digital signatures are: SHA-256, SHA-384, SHA-512, SHA-512/256 or SHA3-256/384/512.  SHA-1 must not be used for digital signatures and other applications that require collision resistance.

**TL;DR**; from security perspective we should be fine continuing with our current RSA setup.

References:

- https://levelup.gitconnected.com/demystifying-ssh-rsa-in-openssh-deprecation-notice-22feb1b52acd
- https://www.rapidsslonline.com/blog/sha1-vs-sha2-vs-sha256-whats-the-difference/

### Elliptic Curve Algorithms

 From security perspective they are comparable to a well-configured RSA setup (>`3096` bit length, `SHA-2` signing).

The benefit of using elliptic curve algorithms is that they work with much smaller random numbers (don't need to be prime). This leads to a performance advantage over RSA.

Product standard security (SEC-266) further states:

- Recommended elliptic curves: For compatibility reasons (browsers) it is required to support the Prime Field Curves listed in NIST Suite B with following length 224, 256, 384 and 521 bit. As NIST curves are partially debated due to lack of design transparency it is furthermore recommended to support respective Brainpool Curves, and edwards25519 (128 bit) / edwards448 (224 bit)  curves. 

#### `ECDS` (NIST)

Some argument that the disadvantage over `ED25519` is that there was less transparency regarding the disgn of the algorithm (curves, seeds) and Snowden publications revealed that NSA was involved in the standardization process. For `ED25519` this is transparently explained and verifyable.

#### `ED25519` / edwards25519

Overall more popular than `ECDS`. Doesn't have the same intransparency issues as `ECDS`.

**TL;DR**: `ED25519` should be a reasonable alternative to our RSA setup. We won't gain relevant security improvements by changing. Also we can't expect significant performance gains. We use the algorithm for key generation which is only relevant when rotating the certs. Since we are talking about our own PKI (SKR to KCP communication with self-signed certs), we also shouldn't need to worry about compatibility issues.


References:

- https://cryptsus.com/blog/how-to-secure-your-ssh-server-with-public-key-elliptic-curve-ed25519-crypto.html
- https://security.stackexchange.com/questions/230708/should-i-be-using-ecdsa-keys-instead-of-rsa-for-ssh

> consider small PoC

Tried to change the cert setup to use `ED25519`without success. Upon sending an event to KCP, skr-webhook fails with:

```txt
2025-01-09T07:58:23Z	ERROR	skr-webhook	1736409495 Post [POST] https://host.k3d.internal:9443/v1/lifecycle-manager/event request-0 retry-2 error: Post "https://host.k3d.internal:9443/v1/lifecycle-manager/event": remote error: tls: certificate required
1736409498 Post [POST] https://host.k3d.internal:9443/v1/lifecycle-manager/event request-0 retry-3 error: Post "https://host.k3d.internal:9443/v1/lifecycle-manager/event": remote error: tls: certificate required
1736409503 Post [POST] https://host.k3d.internal:9443/v1/lifecycle-manager/event request-0 retry-4 error: Post "https://host.k3d.internal:9443/v1/lifecycle-manager/event": remote error: tls: certificate required
	{"postBody": {"owner":{"Namespace":"kcp-system","Name":"kyma-sample"},"watched":{"Namespace":"kyma-system","Name":"default"},"watchedGvk":{"group":"operator.kyma-project.io","version":"v1beta2","kind":"Kyma"}}, "error": "kcp request failed\nPost \"https://host.k3d.internal:9443/v1/lifecycle-manager/event\": remote error: tls: certificate required"}
github.com/kyma-project/runtime-watcher/skr/internal.(*Handler).sendRequestToKcp
	/app/internal/handler.go:271
github.com/kyma-project/runtime-watcher/skr/internal.(*Handler).validateResources
	/app/internal/handler.go:174
github.com/kyma-project/runtime-watcher/skr/internal.(*Handler).Handle
	/app/internal/handler.go:90
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2220
net/http.(*ServeMux).ServeHTTP
	/usr/local/go/src/net/http/server.go:2747
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:3210
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:2092
2025-01-09T07:58:23Z	INFO	skr-webhook	kcp request failed
Post "https://host.k3d.internal:9443/v1/lifecycle-manager/event": remote error: tls: certificate required
2025-01-09T07:58:23Z	INFO	skr-webhook	Preparing response for AdmissionReview: Kyma UPDATE kcp request failed
Post "https://host.k3d.internal:9443/v1/lifecycle-manager/event": remote error: tls: certificate required
```

I have spent some hours trying to find the reason but without success unfortunately. Certificate and Key synced to the SKR look okay. Also the logs don't show any hint about something wrong with the certificate.

**Comment by c-pius** [comment]
> - [ ]  decided whether to stick with RSA or update to a new algorithm with the team and created necessary follow-ups 

I think we are at a point now where we can decide whether we want to update to the new algorithm or not. Personally, I don't see a significant improvement. Since the simple PoC didn't work we also would need to spend some time figuring out what to do exactly.

**Comment by c-pius** [comment]
Decided to stick with RSA for now. We may re-investigate once we have the need to transition to a new algorithm.
